zio-http/src/test/scala/zio/http/TestServer.scala

```diff
@@ -0,0 +1,155 @@
+    val handler: PartialFunction[(State, Request), (State, Response)] = {
+      case (state, realRequest) if {
+            // The way that the Client breaks apart and re-assembles the request prevents a straightforward
+            //    expectedRequest == realRequest
+            expectedRequest.url.relative == realRequest.url &&
+            expectedRequest.method == realRequest.method &&
+            expectedRequest.headers.toSet.forall(expectedHeader => realRequest.headers.toSet.contains(expectedHeader))
+          } =>
+        (state, response)
+    }
+    addHandlerState(handler)
+  }
+
+  /**
+   * Define stateless behavior for the Server.
+   *
+   * @param pf
+   *   The stateless behavior
+   *
+   * @return
+   *   The TestSever with new behavior.
+   *
+   * @example
+   *   {{{
+   *   ZIO.serviceWithZIO[TestServer[Unit]](_.addHandler { case _: Request =>
+   *       Response(Status.Ok)
+   *   }
+   *   }}}
+   */
+  def addHandler(pf: PartialFunction[Request, Response]): TestServer[State] = {
+    val handler: PartialFunction[(State, Request), (State, Response)] = {
+      case (state, request) if pf.isDefinedAt(request) => (state, pf(request))
+    }
+    addHandlerState(handler)
+  }
+
+  /**
+   * Define stateful behavior for Server
+   * @param pf
+   *   Stateful behavior
+   * @return
+   *   The TestSever with new behavior.
+   *
+   * @example
+   *   {{{
+   * ZIO.serviceWithZIO[TestServer[Int]](_.addHandlerState { case (state, _: Request) =>
+   *   if (state > 0)
+   *     (state + 1, Response(Status.InternalServerError))
+   *   else
+   *     (state + 1, Response(Status.Ok))
+   * }
+   *   }}}
+   */
+  def addHandlerState(
+    pf: PartialFunction[(State, Request), (State, Response)],
+  ): TestServer[State] =
+    copy(behavior = behavior.orElse(pf))
+
+  def install(implicit
+    trace: zio.Trace,
+  ): UIO[Unit] =
+    driver.addApp(
+      Http.fromFunctionZIO((request: Request) => state.modify(state1 => behavior((state1, request)).swap)),
+    )
+
+  override def install[R](httpApp: HttpApp[R, Throwable], errorCallback: Option[ErrorCallback])(implicit
+    trace: zio.Trace,
+  ): URIO[R, Unit] =
+    ZIO.environment[R].flatMap { env =>
+      driver.addApp(
+        if (env == ZEnvironment.empty) httpApp.asInstanceOf[HttpApp[Any, Throwable]]
+        else httpApp.provideEnvironment(env),
+      ) *> {
+        val func =
+          (request: Request) => state.modify(state1 => behavior((state1, request)).swap)
+
+        val app: HttpApp[Any, Nothing] = Http.fromFunctionZIO(func)
+        driver.addApp(app)
+      }
+
+    } *> setErrorCallback(errorCallback)
+
+  private def setErrorCallback(errorCallback: Option[ErrorCallback]): UIO[Unit] =
+    driver
+      .setErrorCallback(errorCallback)
+      .unless(errorCallback.isEmpty)
+      .map(_.getOrElse(()))
+
+  override def port: Int = bindPort
+}
+
+object TestServer {
+
+  val make: ZIO[Driver with Scope, Throwable, TestServer[Unit]] =
```

Comment:
It would be more common for this to be called `layer` rather than `make`, I think.

Reply:
Corrected in my current branch